File: coding_skill_cpp/06_factory_pattern.cpp
/* C++ Factory - 템플릿과 unique_ptr */
#include <iostream>
#include <memory>
#include <map>
#include <string>
using namespace std;

class Product {
public:
    virtual ~Product() = default;
    virtual void use() = 0;
};

class ProductA : public Product {
public:
    void use() override { cout << "  → ProductA 사용" << endl; }
};

class ProductB : public Product {
public:
    void use() override { cout << "  → ProductB 사용" << endl; }
};

class Factory {
public:
    static unique_ptr<Product> create(const string& type) {
        if (type == "A") return make_unique<ProductA>();
        if (type == "B") return make_unique<ProductB>();
        return nullptr;
    }
};

// 템플릿 팩토리
template<typename T>
class GenericFactory {
public:
    static unique_ptr<T> create() {
        return make_unique<T>();
    }
};

int main() {
    cout << "\n=== C++ Factory Pattern ===" << endl;
    
    auto productA = Factory::create("A");
    auto productB = Factory::create("B");
    
    productA->use();
    productB->use();
    
    // 템플릿 팩토리
    auto product = GenericFactory<ProductA>::create();
    product->use();
    
    return 0;
}

File: coding_skill_py/06_factory_pattern.py
"""Python Factory - classmethod와 딕셔너리"""
from abc import ABC, abstractmethod
from typing import Dict, Type

class Product(ABC):
    @abstractmethod
    def use(self):
        pass

class ProductA(Product):
    def use(self):
        print("  → ProductA 사용")

class ProductB(Product):
    def use(self):
        print("  → ProductB 사용")

class Factory:
    _products: Dict[str, Type[Product]] = {
        'A': ProductA,
        'B': ProductB
    }
    
    @classmethod
    def create(cls, product_type: str) -> Product:
        if product_type in cls._products:
            return cls._products[product_type]()
        raise ValueError(f"Unknown type: {product_type}")
    
    @classmethod
    def register(cls, name: str, product_class: Type[Product]):
        cls._products[name] = product_class

def main():
    print("\n=== Python Factory Pattern ===")
    
    productA = Factory.create('A')
    productB = Factory.create('B')
    
    productA.use()
    productB.use()

if __name__ == "__main__":
    main()
